### PR TITLE
Add process STAR gene quantification

### PR DESCRIPTION
**Pull request by OtonicarJan**
## Checklist

* [x] Update CHANGELOG.rst for each commit separately:
  * Pay attention to write entries under the "Unreleased" section.
  * Mark all breaking changes as "**BACKWARD INCOMPATIBLE:**" and put them
    before non-breaking changes.
  * If a commit modifies a feature listed under "Unreleased" section,
    it might be sufficient to modify the existing CHANGELOG entry from previous
    commit(s).
* [x] Bump the process version:
  * **MAJOR version (first number)**: Backward incompatible changes (changes
    that break the api/interface). Examples: renaming the input/output, adding
    mandatory input, removing input/output...
  * **MINOR version (middle number)**: add functionality or changes in a
    backwards-compatible manner. Examples: add output field, add non-mandatory
    input parameter, use a different tool that produces same results...
  * **PATCH version (last number)**: changes/bug fixes that do not affect
    the api/interface. Examples: typo fix, change/add warning messages...
* [x] All inputs are used in process.
* [x] All output fields have a value assigned to them.

### Processes ###

* Code that was once accepted should be kept like it is. It should be changed
  only when its functionality has changed. Small parts may be changed
  to fix bugs (by changing code only to make it nicer you might introduce a bug).
* Set the optimal number of cores that the tool can use (check if parallelization
  can be performed).
* Description of the tool should be self explanatory with links to the tool's
  homepage and publication.
* Write explanatory input labels and add short command line parameter names
  in square brackets (for example: Maximum length [``--maxlength``]).
* Where possible provide default values for the process input fields.
  If they are provided, then ``required = False`` is not needed.
* Add descriptive error and warning messages.
* Use ``Cmd()`` for each call in the process (call of a tool, a
  Bash command, a Python script or an R script) and handle return codes.
* Add ``self.progress()`` calls where appropriate.

### Commit and PR messages ###

* Each commit should be minimal (i.e. 1 change) and self-contained (including
  tests).
* Mark incomplete PRs with the [WIP] (Work In Progress) tag.
* Do not paste links to private repositories from the public one.

### Tests ###

* Add meaningful names of test files:
  * Genomes: ``genome_species.fa.gz``
  * Annotations: ``annotation_species.gtf.gz``
  * Adapters: ``adapters-name/type-source``.
* Specific files should be saved in a folder with the name of the tested tool.
  This folder should include separate subfolders for inputs and outputs.
* Keep test files' size small. If there is no other option, add the large
  test file to the ``files/large`` directory so it will be tracked with Git LFS.
* Pay attention to edge cases: what will happen if any data is
  missing, if the length of some list is 0, if a string is given where one
  expects an integer, if a file is empty, ...
* When testing workflows, try to verify workflow success by checking one
  of the last steps. Pick the one that is most likely to fail. Also test
  that all steps have 'OK' status.

### CODE REVIEW

**Review by romunov — CHANGES_REQUESTED** [review]
Overall structure of the process and tests seems fine. There are a few things we should probably address. See inline comments.

File: resolwe_bio/processes/expression/star_quantification.py [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+                    "Re-run analysis in user-selected strandedness mode."
+                )
+        else:
+            strandedness = STRANDEDNESS_CODES[inputs.assay_type]
+
+        raw_counts = "rc.txt"
+        tpm = "tpm.txt"
+        cpm = "cpm.txt"
+
+        features = prepare_gene_counts(
```
[the remark this answers is not in the document]
Comment (OtonicarJan) [reply resolwe_bio/processes/expression/star_quantification.py:424]: `raw_counts` dataframe that is generated with this function is needed for `rnanorm`. I can remove the line `return exp.iloc[4:].index.to_list()` from the function `prepare_gene_counts` and then read the index of `raw_counts` after `rnanorm` to get features?

File: resolwe_bio/processes/expression/star_quantification.py [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+    }
+    scheduling_class = SchedulingClass.BATCH
+
+    class Input:
+        """Input fields."""
+
+        aligned_reads = DataField(
+            "alignment:bam:star",
+            label="Aligned reads",
+            description="If gene counts are not present in the output "
```
[the remark this answers is not in the document]
Comment (OtonicarJan) [reply resolwe_bio/processes/expression/star_quantification.py:201]: Currently, gene counts are not computed by default in STAR aligner process. They will be computed by default in the RNA-seq, but not all alignments need them (globin, rRNA, Qorts).

File: resolwe_bio/processes/expression/star_quantification.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (romunov) [reply resolwe_bio/processes/expression/star_quantification.py:429]: Forgot about the side effects of the function. Perhaps this should be noted in the comment, something along the lines of

```python
# prepare_gene_counts() has a side effect of creating csv files
# used in rnanorm.
features = prepare_gene_counts(...)
```

File: resolwe_bio/processes/expression/star_quantification.py [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+            "of STAR aligner process this process will fail.",
+        )
+
+        annotation = DataField(
+            "annotation",
+            label="Annotation",
+            description="GTF and GFF3 annotation formats are supported",
+        )
+
+        normalization_type = StringField(
```
[the remark this answers is not in the document]
Comment (jkokosar) [reply resolwe_bio/processes/expression/star_quantification.py:216]: Let's leave this in.

File: resolwe_bio/processes/expression/star_quantification.py [hunk 4]
```diff
@@ -0,0 +1,502 @@
...
+    """Rename columns and compress the expression files."""
+    exp = pd.read_csv(
+        infile,
+        sep="\t",
+        usecols=["FEATURE_ID", "EXPRESSION"],
+        index_col="FEATURE_ID",
+        squeeze=True,
+        float_precision="round_trip",
+    )
+    exp.to_csv(
```
[the remark this answers is not in the document]
Comment (jkokosar) [reply resolwe_bio/processes/expression/star_quantification.py:73]: We do unzip the files before doing the assert in tests, so removing time stamps is not necessary.

File: resolwe_bio/processes/expression/star_quantification.py [hunk 5]
```diff
@@ -0,0 +1,509 @@
...
+    requirements = {
+        "expression-engine": "jinja",
+        "executor": {
+            "docker": {
+                "image": "public.ecr.aws/genialis/resolwebio/rnaseq:6.2.0",
+            },
+        },
+        "resources": {
+            "cores": 4,
+            "memory": 32768,
```
Comment (jkokosar) [inline resolwe_bio/processes/expression/star_quantification.py:187]: Drop the default requirements to 2 cores and 16384 memory. More resources are needed in the case of large Salmon index files (containing a full genome decoy), but not with the smaller index files that are typically used.

File: resolwe_bio/processes/expression/star_quantification.py [hunk 6]
```diff
@@ -0,0 +1,509 @@
...
+        }
+
+    with open(file=rc_output, mode="wt") as f:
+        json.dump(exp, f)
+
+    return rc_output
+
+
+class STARQuantification(ProcessBio):
+    """STAR quantification.
```
Comment (jkokosar) [inline resolwe_bio/processes/expression/star_quantification.py:169]: I suggest the following description: `Normalize STAR quantification results`. "STAR quantification" wrongly implies that the full quantification by STAR aligner (while aligning) is done here. Modify the class name accordingly.

**Review by jkokosar — APPROVED** [review]

**Review by romunov — APPROVED** [review]
